src/mono/wasm/runtime/driver.c

```diff
@@ -1347,3 +1348,55 @@ mono_wasm_init_finalizer_thread (void)
 	mono_gc_init_finalizer_thread ();
 #endif
 }
+
+#define I52_ERROR_NONE 0
+#define I52_ERROR_NON_INTEGRAL 1
+#define I52_ERROR_OUT_OF_RANGE 2
+
+#define U52_MAX_VALUE ((1ULL << 53) - 1)
+#define I52_MAX_VALUE ((1LL << 53) - 1)
+#define I52_MIN_VALUE -I52_MAX_VALUE
+
+EMSCRIPTEN_KEEPALIVE double mono_wasm_i52_to_f64 (int64_t *source, int *error) {
+	int64_t value = *source;
+
+	if ((value < I52_MIN_VALUE) || (value > I52_MAX_VALUE)) {
+		*error = I52_ERROR_OUT_OF_RANGE;
+		return NAN;
+	}
+
+	*error = I52_ERROR_NONE;
+	return (double)value;
+}
+
+EMSCRIPTEN_KEEPALIVE double mono_wasm_u52_to_f64 (uint64_t *source, int *error) {
+	uint64_t value = *source;
+
+	if (value > U52_MAX_VALUE) {
+		*error = I52_ERROR_OUT_OF_RANGE;
+		return NAN;
+	}
+
+	*error = I52_ERROR_NONE;
+	return (double)value;
+}
+
+EMSCRIPTEN_KEEPALIVE int mono_wasm_f64_to_u52 (uint64_t *destination, double value) {
+	if ((value < 0) || (value > U52_MAX_VALUE))
+		return I52_ERROR_OUT_OF_RANGE;
+	if (floor(value) != value)
```

Comment:
We are hoping for `f64.floor` wasm instruction, right ?
Alternatively we could do bitmask on ulong. `0x7FE0_0000_0000_0000` if any of those are set it's not integer.

Reply:
I don't love the bitmask because it gets us into weird aliasing territory, but maybe it's worth doing. I think that would also force value to get stored to the native stack instead of living in a wasm register, so who knows if it's actually faster... Aiming for simplicity here. What do you think?